Enhanced Time-Series Data Support with High-Level API

mgyoo86 opened the pull request:
This PR introduces a new high-level API to the `ADIOS2.jl` library that significantly enhances time-series data handling capabilities. The main focus is on streamlining the management of multiple timesteps and simplifying data reading operations.

## Key Features

### New Features:
* **High-Level API for Data Reading**:
  - Introduced the `adios_load` function in `src/extended_highlevel/adios_load.jl`, which provides a user-friendly interface for reading variables from ADIOS files. It supports scalar, array, and time-series data, with options for filtering variables by name or regex patterns and selecting specific steps.
  - Updated the `README.md` to include detailed examples of writing and reading time-series data using the new API.

### Code Enhancements:
* **Variable Handling Improvements**:
  - Modified `adios_put!` in `src/highlevel.jl` to ensure variables are only defined if they do not already exist, avoiding redundant definitions when writing a file with steps. [[1]](diffhunk://#diff-f7e88dbda3743e04fdf88bf8e04c091017d14373494fc2ba943160ff13346163R208-R211) [[2]](diffhunk://#diff-f7e88dbda3743e04fdf88bf8e04c091017d14373494fc2ba943160ff13346163R234-R238)

## Usage Examples

The new API can greatly simplify data analysis workflows:

```julia
# Open a file
file = adios_open_serial("simulation.bp", mode_readRandomAccess)

# Read all steps of all variables
all_data = adios_load(file)

# Read a specific variable at a specific step
temperature_data = adios_load(file, "temperature", 5)

# Read multiple variables at multiple steps
selected_data = adios_load(file, ["temperature", "pressure"], 1:3)

# Find variables by pattern matching
temperature_related = adios_load(file, r"temp.*", 0:9)
```

## Note
The tests for this functionality were primarily conducted using serial reading and writing modes and assume the `BP5` file format. Additional testing with other formats and parallel I/O modes may be needed for comprehensive validation in different environments.

eschnett commented on src/extended_highlevel/adios_load.jl:
```diff
@@ -0,0 +1,338 @@
+
+function adios_load(file::AdiosFile,
+                    varNames::Union{AbstractString,
+                                    AbstractArray{<:AbstractString},Regex})
```
This line (with the trailing `Regex`) is a bit weirdly formatted.

eschnett commented on src/extended_highlevel/adios_load.jl:
```diff
@@ -0,0 +1,338 @@
+
+function adios_load(file::AdiosFile,
+                    varNames::Union{AbstractString,
+                                    AbstractArray{<:AbstractString},Regex},
```
Ditto

eschnett commented:
Thanks for your contribution! This looks good at first glance although I haven't tried it yet.

Some minor comments:

Did you reformat the code with `JuliaFormatter`?

 The documentation style of some functions is different; it doesn't say how to actually call the function or what it returns. Compare e.g.
```
export name
"""
    attr_name = name(attribute::Attribute)
    attr_name::Union{Nothing,String}

Retrieve attribute name.
"""
function name(attribute::Attribute)
    ...
```

eschnett commented:
I see that the CI scripts are outdated. I will update them on the main branch. You will then have to merge these changes into your PR.

williamfgc commented:
Yes, thanks for the great contribution! I just bring awareness that the Python high-level API in ADIOS2 use `write` and `read` to write/return a numpy array. It's up to you and @eschnett what makes more sense. This is great!

eschnett commented:
I updated the CI scripts. Next steps: Merge the main branch into this PR and make CI pass.

eschnett commented:
@williamfgc What is your suggestion, concretely? Should this PR be changed? Or the current behaviour of ADIOS2.jl?

williamfgc commented:
@eschnett I wanted to bring awareness, since I'm not familiar with your use-cases. I'll let you decide if following the Python API or not is the way to go for a high-level interface in ADIOS2.jl .

mgyoo86 commented:
@eschnett 
I merged the master and pushed to this PR.
Could you trigger the CI again?

Regarding the test, I had to make a change to the `basic.jl` test because of a broken test when reading an array of String (["42"]). I put `@test_broken` to it so that the test can pass.

Regarding the formatting, yes, I used the `JuliaFormatter` and it put the line break for a long argument description. Do you want it to be in a single line?
For the general formatting (including the docstring), please feel free to edit it to suit the project's style.

mgyoo86 commented:
@williamfgc Thanks for the feedback!
Actually, I was motivated by the `Matlab` bindings in ADIOS2 (see [here](https://github.com/ornladios/ADIOS2/tree/master/bindings/Matlab)), which support similar high-level APIs, including the `adios_load.m`.
I found the `adios_load.m` is very handy to use in real use cases, so now I'm trying to implement a similar (with extended) feature into the `ADIOS2.jl`.

I think the current `ADIOS2.jl` already has some high-level APIs for writing and reading a single field.
However, there is no capability for dealing with time-series data, which this PR aims to contribute.

williamfgc commented:
@mgyoo86 thanks for the background info!  Makes sense, the Matlab bindings were requested by a TAE (formerly Tri Alpha) use-case and Julia is probably closer. Yes, `RandomAccess` is file only and orthogonal to the streaming case. Thanks for the work, it's great to see use cases. Happy to help.

eschnett commented:
@mgyoo86 Let me know if some of the tests only fail on older Julia versions. Feel free to disable testing Julia 1.6 ... 1.9 in your PR at least for the time being.

mgyoo86 commented:
@eschnett 
First, I reverted the `@test_broke` change in `basic.jl` (which seems to be messed up with my locally installed ADIOS2 libs). 
Second, I disabled the Julia "1.6" version in CI, and then CI passed all other tests (from Julia 1.7 to 1.11).
I don't know why Julia 1.6 did not work.. I installed 1.6 on my local laptop to investigate, but somehow it cannot install the MPI package, so I cannot test further for now.

mgyoo86 commented:
## Updated PR with 3 additional commits

**Summary of changes (1e44ff0 → 38d93f8):**
- **New convenience dispatch**: Added `adios_load(bpPath::AbstractString, args...)` to read data directly from BP file paths (with `.bp` extension). This convenience wrapper automatically opens the file in `mode_readRandomAccess`, reads the data, and closes the file. All functionality of `adios_load(file::AdiosFile, ...)` is available through the same argument patterns.
- **Documentation**: Added "Extended High-Level API" section to the documentation, which includes the docstring for the `adios_load` function
- **Testing**: Refactored test scripts for better readability and broader test coverage

The changes focus on documentation completeness and code organization without affecting core functionality. I believe the PR now passes all CI checks. 
@eschnett Could you please review this PR to see if it's ready to merge?

Thanks for the review!

eschnett commented on src/extended_highlevel/adios_load.jl:
```diff
@@ -0,0 +1,400 @@
+- First:
+  - `file` (::`AdiosFile`): ADIOS file (opened with `mode_readRandomAccess`)
+- Second (Optional):
+  - `varName` (::`AbstarctString`): Single variable name or nothing (reads all)
```
`AbstarctString` -> `AbstractString`

eschnett commented:
Apart from a typo in the documentation this looks good!

mgyoo86 commented:
@eschnett 
Thank you for your prompt review and pointing out the typo, which is fixed now :)
Please let me know if this PR needs any further corrections/improvements before merging.
